block/aggregation.go

```diff
@@ -49,48 +49,92 @@ func (m *Manager) lazyAggregationLoop(ctx context.Context, blockTimer *time.Time
 	lazyTimer := time.NewTimer(0)
 	defer lazyTimer.Stop()
 
+	// Track if we've received transactions but haven't built a block yet
+	pendingTxs := false
+
 	for {
 		select {
 		case <-ctx.Done():
 			return
 
 		case <-lazyTimer.C:
+			m.buildingBlock = false
+			pendingTxs = false
+			m.produceBlockLazy(ctx, "lazy_timer", lazyTimer, blockTimer)
+
 		case <-blockTimer.C:
-		}
+			if pendingTxs {
+				m.buildingBlock = true
+				m.produceBlockLazy(ctx, "tx_collection_complete", lazyTimer, blockTimer)
+				pendingTxs = false
+			} else {
+				// Reset the block timer if no pending transactions
+				blockTimer.Reset(m.config.Node.BlockTime.Duration)
+			}
 
-		// Reset the start time
-		start := time.Now()
+		case <-m.txNotifyCh:
+			if m.buildingBlock {
+				continue
+			}
 
-		// Attempt to publish the block regardless of activity
-		if err := m.publishBlock(ctx); err != nil && ctx.Err() == nil {
-			m.logger.Error("error while publishing block", "error", err)
-		}
+			// Instead of immediately producing a block, mark that we have pending transactions
+			// and let the block timer determine when to actually build the block
+			if !pendingTxs {
+				pendingTxs = true
+				m.logger.Debug("Received transaction notification, waiting for more transactions",
+					"collection_time", m.config.Node.BlockTime.Duration)
+				// Reset the block timer to wait for more transactions
+				blockTimer.Reset(m.config.Node.BlockTime.Duration)
```

Comment:
If we do this, then we wouldn't have a consistent block time right? Like let's say block time is 2 seconds, and we just built a block. Then, txsNotifyCh fires after 1 second, it'll then reset the block timer again to 2 seconds. So a block would be produced after 3 seconds now.